crates/solvers/src/domain/solution.rs

```diff
@@ -17,9 +17,80 @@ pub struct Solution {
 }
 
 impl Solution {
+    /// Returns `self` with a new score.
     pub fn with_score(self, score: Score) -> Self {
         Self { score, ..self }
     }
+
+    /// Returns `self` with eligible interactions internalized using the
+    /// Settlement contract buffers.
+    ///
+    /// Currently, this internalizes all interactions with input/outputs where
+    /// all input tokens are trusted and the settlement contract has sufficient
+    /// balance to cover the output tokens.
+    pub fn with_buffers_internalizations(mut self, tokens: &auction::Tokens) -> Self {
+        let mut used_buffers = HashMap::new();
+        for interaction in self.interactions.iter_mut() {
+            let (inputs, outputs, internalize) = match interaction {
+                Interaction::Liquidity(interaction) => (
+                    slice::from_mut(&mut interaction.input),
+                    slice::from_mut(&mut interaction.output),
+                    &mut interaction.internalize,
+                ),
+                Interaction::Custom(interaction) => (
+                    &mut interaction.inputs[..],
+                    &mut interaction.outputs[..],
+                    &mut interaction.internalize,
+                ),
+            };
+
+            let trusted_inputs = inputs.iter().all(|input| {
+                matches!(
+                    tokens.get(&input.token),
+                    Some(auction::Token { trusted: true, .. })
+                )
+            });
+            if inputs.is_empty() || outputs.is_empty() || !trusted_inputs {
+                continue;
+            }
+
+            let Some(required_buffers) =
+                outputs.iter().try_fold(HashMap::new(), |mut map, output| {
+                    let entry = map.entry(output.token).or_default();
+                    *entry = output.amount.checked_add(*entry)?;
+                    Some(map)
+                })
+            else {
+                continue;
+            };
+            let sufficient_buffers = required_buffers.iter().all(|(address, amount)| {
+                tokens
+                    .get(address)
+                    .and_then(|token| {
+                        let used = used_buffers.get(address).copied().unwrap_or_default();
+                        let total = amount.checked_add(used)?;
+
+                        Some(token.available_balance >= total)
+                    })
+                    .unwrap_or_default()
+            });
+            if !sufficient_buffers {
+                continue;
+            }
```

Comment:
I'm wondering if this logic is easier to understand in the try_fold block above (ie check in each round if the new entry overflows `balance - used` and return `None` if it's the case.

Reply:
Implemented your comment in 9899e04c3198d20922d4295cbd2e25a48703a2f4 - LGTM!